core/pkg/ingress/types.go

```diff
@@ -0,0 +1,288 @@
+	// when an update occurs. This is executed frequently because Ingress
+	// controllers watches changes in:
+	// - Ingresses: main work
+	// - Secrets: referenced from Ingress rules with TLS configured
+	// - ConfigMaps: where the controller reads custom configuration
+	// - Services: referenced from Ingress rules and required to obtain
+	//	 information about ports and annotations
+	// - Endpoints: referenced from Services and what the backend uses
+	//	 to route traffic
+	// Any update to services, endpoints, secrets (only those referenced from Ingress)
+	// and ingress trigger the execution.
+	// Notifications of type Add, Update and Delete:
+	// https://github.com/kubernetes/kubernetes/blob/master/pkg/client/cache/controller.go#L164
+	//
+	// ConfigMap content of --configmap
+	// Configuration returns the translation from Ingress rules containing
+	// information about all the upstreams (service endpoints ) "virtual"
+	// servers (FQDN) and all the locations inside each server. Each
+	// location contains information about all the annotations were configured
+	// https://k8s.io/ingress/core/blob/master/pkg/ingress/types.go#L83
+	// The backend returns the contents of the configuration file or an error
+	// with the reason why was not possible to generate the file.
+	//
+	// The returned configuration is then passed to test, and then to reload
+	// if there is no errors.
+	OnUpdate(*api.ConfigMap, Configuration) ([]byte, error)
+	// BackendDefaults returns the minimum settings required to configure the
+	// communication to endpoints
+	BackendDefaults() defaults.Backend
+	// Info returns information about the ingress controller
+	Info() *BackendInfo
+}
+
+// BackendInfo returns information about the backend.
+// This fields contains information that helps to track issues or to
+// map the running ingress controller to source code
+type BackendInfo struct {
+	// Name returns the name of the backend
+	Name string `json:"names"`
+	// Release returns the running version (semver)
+	Release string `json:"release"`
+	// Build returns information about the git commit
+	Build string `json:"build"`
+	// Repository return information about the git repository
+	Repository string `json:"repository"`
+}
+
+// Configuration holds the definition of all the parts required to describe
+type Configuration struct {
+	// HealthzURL location of the URI that is used to know health state
+	// of the ingress controller
+	HealthzURL string `json:"healthzUrl"`
+	// Backends
+	Backends []*Backend `json:"namespace"`
+	// Servers
+	Servers []*Server `json:"servers"`
+	// TCPEndpoints
+	// +optional
+	TCPEndpoints []*Location `json:"tcpEndpoints,omitempty"`
+	// UPDEndpoints
+	// +optional
+	UPDEndpoints []*Location `json:"udpEndpoints,omitempty"`
+	// PassthroughBackend
+	// +optional
+	PassthroughBackend []*SSLPassthroughBackend `json:"passthroughBackend,omitempty"`
```

Comment:
please explain how this list is different from backends. If it's SNI, please state that in the comment since passthrough is a little confusing (even tcp is "passthrough" right? because we're not decoding the bytes in the connection).